examples/amm/src/contract.rs

```diff
@@ -0,0 +1,117 @@
+#![cfg_attr(target_arch = "wasm32", no_main)]
+
+mod state;
+
+use self::state::Amm;
+use amm::{get_pool_balance, update_pool_balance, AmmError};
+use async_trait::async_trait;
+use linera_sdk::{
+    base::{SessionId, WithContractAbi},
+    ApplicationCallResult, CalleeContext, Contract, ExecutionResult, MessageContext,
+    OperationContext, SessionCallResult, ViewStateStorage,
+};
+
+linera_sdk::contract!(Amm);
+
+impl WithContractAbi for Amm {
+    type Abi = amm::AmmAbi;
+}
+
+#[async_trait]
+impl Contract for Amm {
+    type Error = AmmError;
+    type Storage = ViewStateStorage<Self>;
+
+    async fn initialize(
+        &mut self,
+        _context: &OperationContext,
+        _argument: (),
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        Ok(ExecutionResult::default())
+    }
+
+    async fn execute_operation(
+        &mut self,
+        _context: &OperationContext,
+        operation: Self::Operation,
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        match operation {
+            Self::Operation::Swap {
+                input_token,
+                output_token,
+                amount,
+            } => {
+                let input_pool_balance = get_pool_balance(&self.token_pool, &input_token).await?;
+                let output_pool_balance = get_pool_balance(&self.token_pool, &output_token).await?;
+
+                let output_amount =
+                    amm::calculate_output_amount(amount, input_pool_balance, output_pool_balance)?;
+
+                update_pool_balance(
+                    &mut self.token_pool,
+                    &input_token,
+                    input_pool_balance - amount,
+                )?;
+                update_pool_balance(
+                    &mut self.token_pool,
+                    &output_token,
+                    output_pool_balance + output_amount,
+                )?;
```

Comment:
This should probably interact with the two Fungible Token applications representing the tokens, adding the user's provided tokens to one and removing tokens from the other and sending them to the user.

Out of the top of my head, the formula should be something like:

1. User provides userX tokens of type X (either they are provided in a session, or the AMM must run call the X application to move the tokens from the user into the AMMs).
2. AMM now has poolX tokens (in the application instance of token X) and poolY tokens (in the application instance of token Y).
3. AMM calculates ratio = userX/poolX (after poolX contains the user tokens).
4. AMM sends ratio*poolY tokens to the user.

But this should probably be verified :sweat_smile: